include/solvers/navier_stokes_scratch_data.h

```diff
@@ -839,14 +914,175 @@ class NavierStokesScratchData : public PhysicsScratchDataBase
             fluid_particle_relative_velocity_at_particle_location[particle_i]
               .norm() *
             particle_properties[DEM::CFDDEMProperties::PropertiesIndex::dp] /
-            (kinematic_viscosity + DBL_MIN);
+            (kinematic_viscosity_at_particle_location[particle_i] + DBL_MIN);
         particle_i++;
       }
 
     average_fluid_particle_relative_velocity =
       average_fluid_particle_relative_velocity / particle_i;
   }
 
+  /** @brief Interpolates the filtered phase fraction value at the location of the particles.
+   * The latter values are used in calculating the density and viscosity of the
+   * fluid at the particles' locations when VOF is used.
+   *
+   * @param q_particles Quadrature type object that contains the location of the particles
+   * relative to the cell's frame of reference.
+   *
+   * @param current_filtered_solution The present value of the solution for the filtered phase fraction
+   *
+   * @param dof_handler_vof The DoFHandler of the VOF equations
+   */
+
+  template <typename VectorType>
+  void
+  calculate_filtered_vof_at_particle_location(
+    const Quadrature<dim> &q_particles,
+    const VectorType      &current_filtered_solution,
+    const DoFHandler<dim> &dof_handler_vof)
+  {
+    FEValues<dim> fe_values_vof_local_particles((*this->fe_values_vof).get_fe(),
+                                                q_particles,
+                                                update_values |
+                                                  update_quadrature_points |
+                                                  update_JxW_values);
+    const auto   &vof_cell = typename DoFHandler<dim>::cell_iterator(
+      *(*this->fe_values_vof).get_cell(), &dof_handler_vof);
+
+    filtered_phase_values_at_particle_location.resize(number_of_particles);
+
+    fe_values_vof_local_particles.reinit(vof_cell);
+
+    fe_values_vof_local_particles.get_function_values(
+      current_filtered_solution, filtered_phase_values_at_particle_location);
+  }
+
+  /** @brief Calculates the variables needed to compute the particle fluid interactions
+   * in the vans equations.
+   *
+   * @param previous_solution The solution at the previous time step for the fluid's velocity
+   * and pressure
+   *
+   * @param void_fraction_solution The void fraction value calculated with one of the methods of
+   * the VoidFractionBase class.
+   *
+   * @param particle_handler The particle handler object that stores and manages the
+   *  particles in the simulations
+   *
+   * @param dof_handler The DoFHandler object for the Navier-Stokes problem
+   *
+   * @param void_fraction_dof_handler The DoFHandler object for the void fraction calculation
+   */
+
+  template <typename VectorType>
+  void
+  reinit_particle_fluid_interactions(
+    const typename DoFHandler<dim>::active_cell_iterator & /*cell*/,
+    const VectorType & /*current_solution*/,
```

Comment:
Maybe i misunderstand this, but rihgt now the particle_fluid_interaction is only calculated using the previous solution, not the current one? That's a bit of a surprise no?